File: cpp/148.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* h1, ListNode* h2){
        ListNode *newHead = new ListNode(-1);
        ListNode *t = newHead;
        while(h1 || h2){
            if(h1 && (!h2 || h1->val<=h2->val) ){
                t->next = h1;
                h1 = h1->next;
                t = t->next;
            }
            if(h2 && (!h1 || h2->val<h1->val) ){
                t->next = h2;
                h2 = h2->next;
                t = t->next;
            }
        }
        t->next = NULL;
        return newHead->next;
    }

    ListNode* sortList(ListNode* head) {
        if(!head || !head->next) 
            return head;
        ListNode *slow = head;
        ListNode *fast = head->next;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *head2 = slow->next;
        slow->next = NULL;
        return merge(sortList(head), sortList(head2));
    }
};

File: golang/148/148.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	return sort(head, nil)
}

func sort(head, tail *ListNode) *ListNode {
	if head == nil {
		return head
	}

	if head.Next == tail {
		head.Next = nil
		return head
	}

	slow, fast := head, head
	for fast != tail {
		slow = slow.Next
		fast = fast.Next
		if fast != tail {
			fast = fast.Next
		}
	}

	mid := slow
	return merge(sort(head, mid), sort(mid, tail))
}

func merge(head1, head2 *ListNode) *ListNode {
	dummy := &ListNode{}
	t, t1, t2 := dummy, head1, head2
	for t1 != nil && t2 != nil {
		if t1.Val < t2.Val {
			t.Next = t1
			t1 = t1.Next
		} else {
			t.Next = t2
			t2 = t2.Next
		}

		t = t.Next
	}

	if t1 == nil {
		t.Next = t2
	} else if t2 == nil {
		t.Next = t1
	}

	return dummy.Next
}
